rlay-backend-neo4j/src/lib.rs

```diff
@@ -219,77 +199,131 @@ impl Neo4jBackend {
+                            kind_name: key.clone()
+                        });
                     }
-                    continue;
-                }
-                if let Value::String(_) = value {
-                    add_relationship_value(cid.clone(), key, value);
                 }
             }
+
+            val.insert("relationships".to_string(), serde_json::Value::Array(
+                relationships
+                    .iter()
+                    .map(serde_json::to_value)
+                    .map(|r| r.unwrap())
+                    .collect()
+            ));
+            entity_objects.push(val);
         }
 
-        let mut statement_query = format!(
-            "MERGE (n:RlayEntity {{cid: \"{1}\"}}) SET n:{0}",
-            kind_name, cid
-        );
-        if !values.is_empty() {
-            statement_query.push_str(", ");
-            statement_query.push_str(&values.join(", "));
+        let sub_query_labels = EntityKind::variants().iter().map(|variant| {
+            format!(
+                "FOREACH ( ignore in CASE entity.type WHEN '{}' THEN [1] ELSE [] END | SET n:{} )",
+                variant, variant)
+        })
+        .collect::<Vec<String>>()
+        .join("\n");
+
+        // collect all possible field names
+        let mut all_cid_field_names: Vec<String> = vec![];
+
+        macro_rules! cid_field_names {
+            ($kind:path) => {
+                all_cid_field_names.extend(<$kind>::cid_field_names()
+                    .into_iter()
+                    .map(|field| field.to_owned().to_owned())
+                    .collect::<Vec<String>>());
+            };
         }
 
-        let (mut transaction, _) = client.transaction().begin().await?;
+        rlay_ontology::call_with_entity_kinds!(ALL; cid_field_names!);
 
-        // let mut query = client.query();
-        trace!("NEO4J QUERY: {}", statement_query);
-        transaction.add_statement(Statement::new(statement_query));
-        for relationship in relationships {
-            trace!("NEO4J QUERY: {}", relationship);
-            transaction.add_statement(Statement::new(relationship));
-        }
+        all_cid_field_names.sort();
+        all_cid_field_names.dedup();
+
+        let sub_query_relations = all_cid_field_names
+            .iter()
+            .map(|field_name| {
+                format!(
+                    "FOREACH ( ignore in CASE relationship.kind_name WHEN '{}' THEN [1] ELSE [] END | MERGE (n)-[:{}]->(m) )",
+                    field_name, field_name)
+            })
+            .collect::<Vec<String>>()
+            .join("\n");
+
+        let statement_query_main = format!("
+            UNWIND $entities as entity
+            MERGE (n:RlayEntity {{cid: entity.cid }})
+            SET (CASE WHEN entity.type = 'Annotation' THEN n END).value = entity.value
+            SET (CASE WHEN entity.type = 'DataPropertyAssertion' THEN n END).target = entity.target
```

Comment:
Those lines should also be generated via `::call_with_entities`, as they might also change depending on the ontology crate.